crates/tls/src/tls.rs

```diff
@@ -0,0 +1,217 @@
+            .with_client_cert_verifier(client_verifier) // enforcing mTLS
+            .with_single_cert(vec![p2p_cert.der().clone()], p2p_private_key.clone_key())?,
+    ))
+}
+
+/// Builds a [`rustls::client::ClientConfig`] for peer-to-peer connections.
+///
+/// The returned configuration:
+/// - Restricts protocol versions to `constants::TLS_PROTOCOL_VERSION` (e.g., TLS 1.3 only).
+/// - Trusts servers presenting certificates that chain back to the provided `root_cert_store`
+///   (in this setup, the dummy issuer).
+/// - Configures the client to authenticate itself with `p2p_cert` and `p2p_private_key`,
+///   so that both peers verify each other’s identity (mutual TLS).
+///
+/// # Parameters
+/// - `root_cert_store`: Store of trusted root certificates used to verify server certificates.
+/// - `p2p_cert`: Client certificate representing this node in the P2P network.
+/// - `p2p_private_key`: DER-encoded private key matching the public key contained in `p2p_cert`.
+///
+/// # Returns
+/// A configured [`rustls::client::ClientConfig`] ready for constructing a
+/// [`tokio_rustls::TlsConnector`].
+///
+/// # Errors
+/// Returns an error if the provided client certificate and private key do not form a valid identity.
+fn client_tls_config(
+    root_cert_store: Arc<rustls::RootCertStore>,
+    p2p_cert: &rcgen::Certificate,
+    p2p_private_key: &PrivateKeyDer,
+) -> anyhow::Result<Arc<rustls::client::ClientConfig>> {
+    Ok(Arc::new(
+        rustls::ClientConfig::builder_with_protocol_versions(&[constants::TLS_PROTOCOL_VERSION])
+            .with_root_certificates(root_cert_store)
+            .with_client_auth_cert(vec![p2p_cert.der().clone()], p2p_private_key.clone_key())?,
+    ))
+}
+
+/// Builds both server and client TLS configurations for a P2P node.
+///
+/// The configuration process:
+/// - Generates a dummy self-signed certificate authority and adds it to the trusted root store.
+/// - Derives a keypair from the provided `p2p_private_key`.
+/// - Issues a peer certificate for that keypair, signed by the dummy authority.
+/// - Converts the private key into DER format for use with rustls.
+/// - Constructs:
+///   - a [`rustls::server::ServerConfig`] that requires client certificates chaining to the dummy root (mTLS),
+///   - a [`rustls::client::ClientConfig`] that presents the issued peer certificate and trusts the same root.
+///
+/// # Parameters
+/// - `p2p_private_key`: The Ed25519 secret key of this node, used as its identity in P2P handshakes.
+///
+/// # Returns
+/// A tuple `(Arc<ServerConfig>, Arc<ClientConfig>)` containing the TLS configurations
+/// for server and client roles.
+///
+/// # Errors
+/// Returns an error if certificate creation, key conversion, or TLS configuration fails.
+pub fn configure_tls(
+    p2p_private_key: &ed25519_dalek::SigningKey,
+) -> anyhow::Result<(
+    Arc<rustls::server::ServerConfig>,
+    Arc<rustls::client::ClientConfig>,
+)> {
```

Comment:
suggestion: While I see the original signature constructs Arcs directly, I find that weird. This library should not make decisions of which smart pointers (if any) downstream applications chose to use. From what I can see, there's no gain in creating the Arcs in this function as opposed to creating them later, when they are actually needed.